### ISSUE TITLE
MSVC compiler suggested options include /Ox but not /O2 when /O2 is what a user would most likely want

### ISSUE THREAD

**Issue by angusgraham** [compiler-explorer/compiler-explorer#2065]
![image](https://user-images.githubusercontent.com/11641291/86676473-816fef00-bfaf-11ea-9745-70d0c298f775.png)

Hopefully you can see in the attached image, the suggested options for the MSVC compiler include /O1 /Og /Ox and /Od but not /O2 which is offered in VS's UI.

Here's what the MS docs have to say about the misleadingly named "Full Optimization" /Ox:
https://docs.microsoft.com/en-us/cpp/build/reference/ox-full-optimization?view=vs-2019
-------------------------
/Ox (Enable Most Speed Optimizations)

The /Ox compiler option enables a combination of optimizations that favor speed. In some versions of the Visual Studio IDE and the compiler help message, this is called full optimization, but the /Ox compiler option enables only a subset of the speed optimization options enabled by /O2.
-------------------------

I think a user would be most interested in learning what Visual Studio would put out with the full suite of speed optimizations and not just the subset that /Ox provides, so if one of the two is going to be omitted from the list it should be /Ox and not /O2.

**Comment by partouf** [comment]
This is interesting info, thanks.

Though the options are not really chosen by us as you might think.

They are usually/sometimes based on popular arguments used by users in the past years.
And if that information is not available, we fill them whatever the command line help offers and contains something about optimization or std version. (https://github.com/compiler-explorer/compiler-explorer/blob/master/lib/compiler-arguments.js)

If you are looking at the MS compilers, you are looking at the latter option, so just filled from the help. The Wine powered MSVC compilers that we host are usually filled from popular arguments so that does include /O2

I must note btw, that the descriptions of /Ox and /O2 differ from version to version and seem to be the opposite in VS2015 for example from what it is now in VS2019. So I'm not really sure how to approach this problem unless we want to offer preselected options per compiler, but I'm not really in favor of that given the number of compilers that we host.

**Comment by angusgraham** [comment]
Yeah my sense is that /Ox used to be called "Full Optimization" because it combined different kinds of optimizations like inlining and frame pointer omission.
Then later MS introduced the concept of optimizing for speed or size but /Ox kept its old name for reasons of inertia or legacy.
I seem to remember there was a time when there was an option that was a blend of speed/size optimizations - so that may have been /Ox's name at one point.
Then more recently /Ox had the reference to size removed from its name size it doesn't actually do any size optimizations (?), and also had the word "Full" removed from its name since that's misleading.
Not sure if that's all correct, just guesses.

**Comment by corob-msft** [comment]
It's true that the Microsoft-recommended optimization flags are now [/Od (Disable optimization)](https://docs.microsoft.com/en-us/cpp/build/reference/od-disable-debug) for debug builds, or [/O1 (Favor size)](https://docs.microsoft.com/en-us/cpp/build/reference/o1-o2-minimize-size-maximize-speed) or [/O2 (Favor speed)](https://docs.microsoft.com/en-us/cpp/build/reference/o1-o2-minimize-size-maximize-speed) for release builds.

The /O2 option is a shorthand way of specifying the /Ob2, /Oy (on x86), /Oi, /Ot, /GF, and /Gy optimization flags. The [/Ox](https://docs.microsoft.com/en-us/cpp/build/reference/ox-full-optimization) option is a strict subset of /O2 which removes /GF (Eliminate duplicate strings) and /Gy (Enable function-level linking). The latter two flags didn't exist when /Ox was created as a shorthand way of specifying all the available optimizations for speed, or "full optimization." The "full optimization" name stuck around confusing people long after the additional optimizations became available and were added to both /O1 and /O2, but not to /Ox. We didn't update the docs until late 2017, but the change was effective much earlier, including in all MSVC compiler versions in current support.

The [/Og](https://docs.microsoft.com/en-us/cpp/build/reference/og-global-optimizations) flag is deprecated because it doesn't actually enable any additional optimizations. All of the optimizations it used to enable are now on by default in any optimization mode except /Od. This has also been true for some time; it's true in all MSVC compiler versions in current support (roughly, from VS2010 on).

Hope this is useful for decision making. I'm one of the [docs.microsoft.com/cpp](https://docs.microsoft.com/cpp) content maintainers; feel free to reach out to me for info or clarifications on our docs.

**Comment by angusgraham** [comment]
Thanks corob. I've noticed that /Og is still referenced throughout the cpp docs for example the docs for /Ot,/Os say "If you use **/Os** or **/Ot**, then you must also specify [/Og](og-global-optimizations.md) to optimize the code" but that is no longer correct.
I've submitted a pr for that page:
https://github.com/MicrosoftDocs/cpp-docs/pull/2300
But as I said in a comment to that pr:
Looking at a grep of cpp-docs for /Og (https://github.com/MicrosoftDocs/cpp-docs/search?q=Og&unscoped_q=Og) it looks like most references to /Og are similarly misleading. I'm not going to fix them all but I suggest someone should.

**Comment by partouf** [comment]
The more info the more excellent! but how can a computer interpret this excellence, that is the question.
So if we do /? and interpret the lines -> how can we make sure /O2 reaches the top? (older versions are fine because of learning curve)

**Comment by angusgraham** [comment]
I don't know the answer, but at least it seems like the problem will slowly fix itself as /Ox loses favour to /O2 over time.

**Comment by corob-msft** [comment]
It doesn't get any closer to the top than this, from the current MSVC compiler:
```
C:\gh\corob-msft\cpp-docs>cl /nologo /?
                         C/C++ COMPILER OPTIONS


                              -OPTIMIZATION-

/O1 maximum optimizations (favor space) /O2 maximum optimizations (favor speed)
/Ob<n> inline expansion (default n=0)   /Od disable optimizations (default)
/Og enable global optimization          /Oi[-] enable intrinsic functions
/Os favor code space                    /Ot favor code speed
/Ox optimizations (favor speed)         /Oy[-] enable frame pointer omission
/favor:<blend|ATOM> select processor to optimize for, one of:
    blend - a combination of optimizations for several different x86 processors
    ATOM - Intel(R) Atom(TM) processors
```

/O1 and /O2 have been there at the top since at least VS 2013, which is the oldest version on my machine, but /Ox was misleadingly called "maximum optimizations" through VS 2015. :(

**Comment by angusgraham** [comment]
I think he means reaches the top in the corpus of user-chosen args that godbolt's learning algo uses to make suggestions. Perhaps it is possible to "poison" the corpus that the algo uses be removing references to /Ox and adding references to /O2 directly in that database?

**Comment by corob-msft** [comment]
Ah, that's pretty cool, I didn't know that was how it worked. But yes, anything to get the /O2 word out is good.

**Comment by partouf** [comment]
right its more about that currently the sorting is based on keywords like 'optimization', but maybe 'maximum' should count +1  and speed +1 ?

**Comment by partouf** [comment]
> right its more about that currently the sorting is based on keywords like 'optimization', but maybe 'maximum' should count +1 and speed +1 ? https://github.com/compiler-explorer/compiler-explorer/blob/master/lib/compiler-arguments.js#L94 (again, wine compilers should be influenced by user preference, but MSVC instances will notice changes in theree)

**Comment by partouf** [comment]
> @partouf Is it possible just to display all of the help there?

No, there's just not enough UI space, but we did recently add this https://github.com/compiler-explorer/compiler-explorer/pull/3597
Which adds the possibility for you to make custom json files with only the important compiler flags. This still has a bit of magic (perhaps too much), but it should be possible to make /O2 the top result.
